src/JustSaying/Messaging/Channels/RoundRobinQueueMultiplexer.cs

```diff
@@ -0,0 +1,142 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using JustSaying.Messaging.Monitoring;
+using Microsoft.Extensions.Logging;
+
+namespace JustSaying.Messaging.Channels
+{
+    internal sealed class RoundRobinQueueMultiplexer : IMultiplexer, IDisposable
+    {
+        private readonly IList<ChannelReader<IQueueMessageContext>> _readers;
+        private Channel<IQueueMessageContext> _targetChannel;
+
+        private readonly SemaphoreSlim _readersLock = new SemaphoreSlim(1, 1);
+        private readonly SemaphoreSlim _startLock = new SemaphoreSlim(1, 1);
+        readonly ILogger<RoundRobinQueueMultiplexer> _logger;
+
+        private bool _started = false;
+        private int _channelCapacity;
+
+        public Task Completion { get; private set; }
+
+        public RoundRobinQueueMultiplexer(ILoggerFactory loggerFactory)
+        {
+            _readers = new List<ChannelReader<IQueueMessageContext>>();
+            _logger = loggerFactory.CreateLogger<RoundRobinQueueMultiplexer>();
+
+            // TODO: make configurable
+            _channelCapacity = 100;
+            _targetChannel = Channel.CreateBounded<IQueueMessageContext>(_channelCapacity);
+        }
+
+        public void ReadFrom(ChannelReader<IQueueMessageContext> reader)
+        {
+            if (reader == null) throw new ArgumentNullException(nameof(reader));
+
+            _readersLock.Wait();
+            try
+            {
+                _readers.Add(reader);
+            }
+            finally
+            {
+                _readersLock.Release();
+            }
+
+            reader.Completion.ContinueWith(c => RemoveReader(reader), TaskScheduler.Default);
```

Comment:
```suggestion
        async Task OnReaderCompletion()
        {
            await reader.Completion.ConfigureAwait(false);
            RemoveReader(reader);
        }

        _ = OnReaderCompletion();
```

https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#prefer-await-over-continuewith

Reply:
Can't say I like it, but I understand it 😬

Reply:
Hah, also would this be a good thing? 😁
`ContinueWith also does not capture the SynchronizationContext and as a result is actually semantically different to async/await`

Reply:
Yes, and here `ContinueWith` actually fits well, it's just part of a more general set of guidance to treat `ContinueWith` as obsolete, perhaps that's a bit dogmatic. Elsewhere in JustSaying we have opted to follow the guidance even when "we've known better".

Reply:
That's fair, thanks for the background. Will keep this in mind elsewhere 👍